libkbfs/prefetcher_test.go

```diff
@@ -1066,9 +1077,401 @@ func TestSyncBlockCacheWithPrefetcher(t *testing.T) {
+	testPrefetcherCheckGet(t, config.BlockCache(), aPtr, a, FinishedPrefetch,
+		TransientEntry)
+
+	// Then we wait for the pending prefetches to complete.
+	waitForPrefetchOrBust(t, q.Prefetcher().Shutdown())
+}
+
+func TestPrefetcherUnsyncedPrefetchChildEvictedCanceled(t *testing.T) {
+	t.Log("Partial regression test for KBFS-2588: when a prefetched block " +
+		"has children waiting on a prefetch, and it is canceled, subsequent " +
+		"attempts to prefetch that parent block panic.")
+	// Note: this test actually passes, because as long as the block count of
+	// the parent is non-zero, the first child that completes it will also
+	// remove it from the tree. See
+	// TestPrefetcherUnsyncedPrefetchRootEvictedCanceled for the actual
+	// regression test that reproduces the panic.
+	dbc, dbcConfig := initDiskBlockCacheTest(t)
+	q, bg, config := initPrefetcherTestWithDiskCache(t, dbc)
+	bcache := config.testCache
+	defer shutdownPrefetcherTest(q)
+	ctx := context.Background()
+	kmd := makeKMD()
+	prefetchSyncCh := make(chan struct{})
+	q.TogglePrefetcher(true, prefetchSyncCh)
+	notifySyncCh(t, prefetchSyncCh)
+
+	t.Log("Initialize a folder tree with structure: " +
+		"root -> {a, b}")
+	rootPtr := makeRandomBlockPointer(t)
+	root := &DirBlock{Children: map[string]DirEntry{
+		"a": makeRandomDirEntry(t, File, 10, "a"),
+		"b": makeRandomDirEntry(t, File, 10, "b"),
+	}}
+	aPtr := root.Children["a"].BlockPointer
+	bPtr := root.Children["b"].BlockPointer
+	a := makeFakeFileBlock(t, true)
+	b := makeFakeFileBlock(t, true)
+
+	encRoot, serverHalfRoot :=
+		setupRealBlockForDiskCache(t, rootPtr, root, dbcConfig)
+	encA, serverHalfA := setupRealBlockForDiskCache(t, aPtr, a, dbcConfig)
+	encB, serverHalfB := setupRealBlockForDiskCache(t, bPtr, b, dbcConfig)
+
+	_, _ = bg.setBlockToReturn(rootPtr, root)
+	_, _ = bg.setBlockToReturn(aPtr, a)
+	_, _ = bg.setBlockToReturn(bPtr, b)
+	err := dbc.Put(ctx, kmd.TlfID(), rootPtr.ID, encRoot, serverHalfRoot)
+	require.NoError(t, err)
+	err = dbc.Put(ctx, kmd.TlfID(), aPtr.ID, encA, serverHalfA)
+	require.NoError(t, err)
+	err = dbc.Put(ctx, kmd.TlfID(), bPtr.ID, encB, serverHalfB)
+	require.NoError(t, err)
+
+	t.Log("Fetch dir root.")
+	var block Block = &DirBlock{}
+	ch := q.Request(context.Background(), defaultOnDemandRequestPriority, kmd,
+		rootPtr, block, TransientEntry)
+	notifySyncCh(t, prefetchSyncCh)
+	err = <-ch
+	require.NoError(t, err)
+	// Notify sync channel for the child blocks `a` and `b`.
+	notifySyncCh(t, prefetchSyncCh)
+	notifySyncCh(t, prefetchSyncCh)
+
+	t.Log("Cancel the prefetch for block 'a'.")
+	q.Prefetcher().CancelPrefetch(aPtr.ID)
+	// Notify sync channel for the cancelation.
+	notifySyncCh(t, prefetchSyncCh)
+
+	t.Log("Set the metadata of the root block in the disk cache to " +
+		"NoPrefetch, simulating an eviction and a BlockServer.Get.")
+	err = dbc.UpdateMetadata(ctx, rootPtr.ID, NoPrefetch)
+	require.NoError(t, err)
+
+	t.Log("Evict the root block from the block cache.")
```

Comment:
The test name contains "ChildEvicted", but we're evicting the root here.  I assume that's intentional, but maybe the test name needs to be clarified?  (Maybe "Evicted" should just be removed from the name?)